pilot/cmd/pilot-agent/status/server.go

```diff
@@ -756,6 +751,31 @@ func (s *Server) handleAppProbeHTTPGet(w http.ResponseWriter, req *http.Request,
 	w.WriteHeader(response.StatusCode)
 }
 
+// Host can either come from user explicitly setting the host in `httpHeaders`, or implicit defined to be IP:port.
+//
+// `httpHeaders`, in an edge case, can set Host multiple times due to header casing.
+// On older Kubernetes versions, this results in 2 Host headers, which Go (pilot-agent) rejects.
+// On newer versions (https://github.com/kubernetes/kubernetes/pull/117323), the first one is picked.
+//
+// Using the implicit hostname also has issues, though. The port will be seen as 15020 by Kubernetes. To handle this, we
+// need to overwrite it
+func determineHost(req *http.Request, prober *Prober) string {
+	for _, header := range prober.HTTPGet.HTTPHeaders {
+		if strings.EqualFold(header.Name, "Host") {
+			// Case 1: explicit set host.
+			// Case insensitive match with the first match winning
+			return header.Value
+		}
+	}
+	if host, _, err := net.SplitHostPort(req.Host); err == nil {
+		// Injected Prober already has port rewritten to a number, so IntValue is always correct
+		port := strconv.Itoa(prober.HTTPGet.Port.IntValue())
+		return net.JoinHostPort(host, port)
```

Comment:
@howardjohn I should be more explicit:, with the example below, I think the host will be rewrite to foo.bar.com:80, while the user expected foo.bar.com:8080

```
   readinessProbe:
          httpGet:
            httpHeaders:
            - name: Host
              value: foo.bar.com:8080
            - name: host
              value: bar
            path: /healthz
            port: 80
```

Reply:
I agree it should be 8080. L767 handles this - I tested this case on both of the PRs and they both handle it correctly i think

Reply:
L767 cannot handle this, go long http removes the `Host` header

Reply:
Oh, sorry. It is reading not the request header but the httpGet.httpHeaders. So lg too

Reply:
Yep. I'll approve yours then?